File: src/main/java/komu/blunt/asm/opcodes/OpPushLabel.java
package komu.blunt.asm.opcodes;

import komu.blunt.asm.Label;
import komu.blunt.asm.Register;
import komu.blunt.asm.VM;

import static com.google.common.base.Preconditions.checkNotNull;

public final class OpPushLabel extends OpCode {
    
    private final Label label;

    public OpPushLabel(Label label) {
        this.label = checkNotNull(label);
    }

    @Override
    public void execute(VM vm) {
        vm.push(label.getAddress());
    }

    @Override
    public boolean modifies(Register register) {
        return false;
    }

    @Override
    public String toString() {
        return String.format("(push (label %s))", label);
    }
}

File: src/main/kotlin/komu/blunt/asm/opcodes/OpPushLabel.kt
package komu.blunt.asm.opcodes

import komu.blunt.asm.Label
import komu.blunt.asm.Register
import komu.blunt.asm.VM

class OpPushLabel(private val label: Label) : OpCode() {

    override fun execute(vm: VM) {
        vm.push(label.address)
    }

    override fun modifies(register: Register) = false

    override fun toString() = "push (label $label)"
}
